**What was tried:** QA workflow run [#25593925269](https://github.com/svv2014/loop/actions/runs/25593925269) on PR #243 (branch `fix/issue-235-loop`).

**What failed:**
```
No structured test names found
```

<details><summary>Last 20 lines of failure output</summary>

```
qa-merge	Run tests	2026-05-09T06:20:25.2627435Z ok 341 validate: valid minimal workflow passes
qa-merge	Run tests	2026-05-09T06:20:25.4083609Z ok 342 validate: all committed starter workflows pass validation
qa-merge	Run tests	2026-05-09T06:20:25.4730119Z ok 343 validate: missing version field fails
qa-merge	Run tests	2026-05-09T06:20:25.5372849Z ok 344 validate: version=2 fails
qa-merge	Run tests	2026-05-09T06:20:25.6016066Z ok 345 validate: name mismatch with filename fails
qa-merge	Run tests	2026-05-09T06:20:25.6668298Z ok 346 validate: duplicate trigger_label within issue_stages fails
qa-merge	Run tests	2026-05-09T06:20:25.7317137Z ok 347 validate: same trigger_label across issue_stages and pr_stages is allowed
qa-merge	Run tests	2026-05-09T06:20:25.7957952Z ok 348 validate: unknown transition target fails
qa-merge	Run tests	2026-05-09T06:20:25.8593746Z ok 349 validate: transition to terminal label 'done' passes
qa-merge	Run tests	2026-05-09T06:20:25.9224008Z ok 350 validate: decisions approve/reject referencing unknown labels fails
qa-merge	Run tests	2026-05-09T06:20:25.9857072Z ok 351 validate: empty workflow with no stages fails
qa-merge	Run tests	2026-05-09T06:20:26.0491233Z ok 352 validate: stage missing trigger_label fails
qa-merge	Run tests	2026-05-09T06:20:26.1127776Z ok 353 validate: stage missing handler fails
qa-merge	Run tests	2026-05-09T06:20:26.1762147Z ok 354 validate: missing handler script produces warning but passes
qa-merge	Run tests	2026-05-09T06:20:26.3481245Z ok 355 default.yaml: every referenced label is canonical
qa-merge	Run tests	2026-05-09T06:20:26.4227284Z ok 356 minimal.yaml: every referenced label is canonical
qa-merge	Run tests	2026-05-09T06:20:26.5184097Z ok 357 docs-only.yaml: every referenced label is canonical
qa-merge	Run tests	2026-05-09T06:20:26.6661578Z ok 358 all shipped workflows pass loop_workflow_validate
qa-merge	Run tests	2026-05-09T06:20:26.9131336Z ok 359 no deprecated alias appears as a workflow label
qa-merge	Run tests	2026-05-09T06:20:26.9462753Z ##[error]Process completed with exit code 1.
```
</details>

**Options:**
(a) Fix the failing test(s) — update the code or test to make them pass, then push a new commit to this PR.
(b) Skip the test(s) — if the test is flaky or irrelevant to this change, annotate it with the appropriate skip marker and document why.

--- a/scripts/po-handler.sh
+++ b/scripts/po-handler.sh
@@ -456,10 +456,11 @@ if loop_run_agent "$TASK_PROMPT" "$ROOT" 2>&1 | tee -a "$LOG_FILE" | tee "$_RUN_
     retry_clear
     transient_clear
     backend_remove_label "$REPO" "$ISSUE_NUM" in-progress
-    # Belt-and-braces: if agent forgot to apply progression label, add 'dev' as safe default.
-    if ! backend_issue_has_any_label "$REPO" "$ISSUE_NUM" dev needs-clarification blocked tracker 'done'; then
-        log "WARN: issue #$ISSUE_NUM has no progression label after PO agent — adding 'dev'"
-        backend_add_label "$REPO" "$ISSUE_NUM" dev
+    # Belt-and-braces: if agent forgot to apply progression label, add dev-queue label as safe default.
+    if ! backend_issue_has_any_label "$REPO" "$ISSUE_NUM" dev needs-dev needs-clarification blocked tracker 'done'; then
+        _fallback_dev_label=$(loop_label_for "$SLUG" "dev")
+        log "WARN: issue #$ISSUE_NUM has no progression label after PO agent — adding '$_fallback_dev_label'"
+        backend_add_label "$REPO" "$ISSUE_NUM" "$_fallback_dev_label"
     fi
 else
     _agent_rc=$?

--- a/tests/po-belt-and-braces.bats
+++ b/tests/po-belt-and-braces.bats
@@ -1,0 +1,114 @@
+#!/usr/bin/env bats
+# tests/po-belt-and-braces.bats — unit tests for po-handler belt-and-braces guard.
+#
+# Verifies that the guard after a successful PO run does not double-label an
+# issue that already carries a canonical dev-queue label (needs-dev, dev, etc.)
+# and that the fallback label respects per-project overrides via loop_label_for.
+
+setup() {
+    REPO_ROOT="$(cd "$BATS_TEST_DIRNAME/.." && pwd)"
+    export LOOP_ROOT="$REPO_ROOT"
+    export LOOP_WORKFLOW_DIR="$REPO_ROOT/config/workflows"
+    # shellcheck source=../lib/workflow.sh
+    source "$REPO_ROOT/lib/workflow.sh"
+
+    cat > "$BATS_TMPDIR/fixture.yaml" <<'YAML'
+version: 1
+projects:
+  - slug: test-proj
+    name: Test Project
+    repo: owner/test-repo
+    root: /tmp/fake
+    default_branch: main
+    workflow: default
+YAML
+    export LOOP_CONFIG="$BATS_TMPDIR/fixture.yaml"
+}
+
+teardown() {
+    rm -f "$BATS_TMPDIR/fixture.yaml" "$BATS_TMPDIR/ops.log"
+}
+
+# Inline the guard logic from po-handler.sh so the test does not need to
+# source the full handler (which pulls in env/backends/runner).
+_run_guard() {
+    local present_labels="$1"   # space-separated labels already on the issue
+    local ops_log="$2"
+    local slug="${3:-test-proj}"
+
+    backend_issue_has_any_label() {
+        # $1=repo $2=num, remaining = labels to check
+        shift 2
+        for lbl in "$@"; do
+            # shellcheck disable=SC2076
+            if [[ " $present_labels " =~ " $lbl " ]]; then
+                return 0
+            fi
+        done
+        return 1
+    }
+
+    backend_add_label() {
+        echo "add $3" >> "$ops_log"
+    }
+
+    log() { :; }
+
+    # Replicate the guard block from po-handler.sh
+    if ! backend_issue_has_any_label "owner/test-repo" "42" dev needs-dev needs-clarification blocked tracker 'done'; then
+        _fallback_dev_label=$(loop_label_for "$slug" "dev")
+        log "WARN fallback"
+        backend_add_label "owner/test-repo" "42" "$_fallback_dev_label"
+    fi
+}
+
+# ---------------------------------------------------------------------------
+
+@test "guard does NOT add dev when needs-dev already present (canonical-vocab project)" {
+    ops_log="$BATS_TMPDIR/ops.log"
+    rm -f "$ops_log"
+
+    _run_guard "needs-dev" "$ops_log"
+
+    # No label should have been added
+    [ ! -f "$ops_log" ] || [ "$(wc -l < "$ops_log")" -eq 0 ]
+}
+
+@test "guard does NOT add dev when legacy dev already present" {
+    ops_log="$BATS_TMPDIR/ops.log"
+    rm -f "$ops_log"
+
+    _run_guard "dev" "$ops_log"
+
+    [ ! -f "$ops_log" ] || [ "$(wc -l < "$ops_log")" -eq 0 ]
+}
+
+@test "guard does NOT add dev when needs-clarification already present" {
+    ops_log="$BATS_TMPDIR/ops.log"
+    rm -f "$ops_log"
+
+    _run_guard "needs-clarification" "$ops_log"
+
+    [ ! -f "$ops_log" ] || [ "$(wc -l < "$ops_log")" -eq 0 ]
+}
+
+@test "guard DOES add fallback dev-queue label when no progression label present" {
+    ops_log="$BATS_TMPDIR/ops.log"
+    rm -f "$ops_log"
+
+    _run_guard "in-progress" "$ops_log"
+
+    grep -q "^add " "$ops_log"
+}
+
+@test "fallback label respects loop_label_for override (canonical project returns needs-dev)" {
+    ops_log="$BATS_TMPDIR/ops.log"
+    rm -f "$ops_log"
+
+    # Override loop_label_for to simulate a canonical-vocab project
+    loop_label_for() { echo "needs-dev"; }
+
+    _run_guard "in-progress" "$ops_log"
+
+    grep -q "^add needs-dev$" "$ops_log"
+}

